packages/snap/src/state.ts

```diff
@@ -0,0 +1,367 @@
+import { Chains, Chain, CosmosAddress } from "./types/chains";
+import { Addresses, Address } from "./types/address";
+import { DirectSecp256k1Wallet } from "@cosmjs/proto-signing";
+
+/**
+ * ChainState is the class to manage all Chain state within Metamask.
+ */
+export class ChainState {
+  /**
+   * Get the chain ids bech32 addresses for all chains in state.
+   *
+   * @returns The bech32 prefixed addresses in a list.
+   * @throws If an error occurs.
+   */
+  public static async getChainAddresses(): Promise<CosmosAddress[]> {
+    let chains = await this.getChains();
+
+    let addressesP = chains.chains.map(async item => {
+      return this.getChainAddress(item.chain_id)
+    })
+    let addresses = await Promise.all(addressesP);
+
+    return addresses
+  }
+  /**
+   * Get the chain ids bech32 address.
+   *
+   * @returns The bech32 prefixed address.
+   * @throws If an error occurs.
+   */
+  public static async getChainAddress(chain_id: string): Promise<CosmosAddress> {
+    let chain = await this.getChain(chain_id);
+    if (chain == null) {
+      throw new Error(`Chain with Chain Id ${chain_id} does not exist.`)
+    }
+
+    // get signer info
+    let node = await snap.request({
+      method: "snap_getBip44Entropy",
+      params: {
+        coinType: typeof chain.slip44 == "number" ? chain.slip44 : 118,
+      },
+    });
+
+    if (typeof node.privateKey === "undefined") {
+      throw Error("Private key from node is undefined");
+    }
+
+    // Create bytes key
+    let pk = node.privateKey;
+    if (pk.startsWith("0x")) {
+      pk = pk.substring(2);
+    }
+
+    // create the wallet
+    let wallet = await DirectSecp256k1Wallet.fromKey(
+      Uint8Array.from(Buffer.from(pk, "hex")),
+      chain.bech32_prefix
+    );
+
+    let address = (await wallet.getAccounts())[0].address;
+
+    return {
+      address,
+      chain_id
+    };
+  }
+  /**
+   * Gets all Cosmos chains from Metamask snap state.
+   *
+   * @returns All chains in Metamask snap state.
+   * @throws If an error occurs.
+   */
+  public static async getChains(): Promise<Chains> {
+    let data = await snap.request({
+      method: "snap_manageState",
+      params: { operation: "get" },
+    });
+    if (data?.chains == undefined || data?.chains == null) {
```

Comment:
`== null` is also true for `undefined`:
```suggestion
    if (data?.chains == null) {
```